### DISCUSSION TITLE
How does Pillow implement Transform.QUAD and Transform.MESH

### DISCUSSION THREAD

**Discussion by kyrollosyanny** [python-pillow/Pillow#8449] [Q&A] [answered]
Hello, 

I am trying to understand how Pillow implements the ```Transform.MESH``` method used in ```ImageOps.deform```. In the source code I found the following 
```python
        if method == MESH:
            # list of quads
            for box, quad in data:
                im.__transformer(box, self, QUAD, quad, resample, fill)
        else:
            im.__transformer((0, 0)+size, self, method, data, resample, fill)
```

So for each entry in the list of quads, it calls the ```__transformer``` function with ```method=QUAD```. Then in the ```__transformer``` function it performs some math on the quad and box data and calls ```self.im.transform2(box, image.im, method, data, resample, fill)``` which I don't seem to find a function for or understand how it does the QUAD transform. 

Any documentation or pointers to source good will be very helpful. 

Thank you

**Comment by Yay295** [comment, marked as answer]
`transform2` is written in C, and [was renamed to `transform`](https://github.com/python-pillow/Pillow/pull/8113) a few months ago.
https://github.com/python-pillow/Pillow/blob/7097a9a3b869f2cf8d335aeb1a6a77c1b31378dc/src/_imaging.c#L3630
https://github.com/python-pillow/Pillow/blob/7097a9a3b869f2cf8d335aeb1a6a77c1b31378dc/src/_imaging.c#L2010-L2072

**Reply by kyrollosyanny** [reply]
Got it! Thank you so much
